File: 155. Min Stack: O(1).rb
# 155. Min Stack
# https://leetcode.com/problems/min-stack/description/

# Stack for push, pop, top and minimum element in O(1) time:
class MinStack
    def initialize()
        # One array for tracking stack and other for minimum value at each step:
        @stack = []
        @min_stack = []
    end

=begin
    :type val: Integer
    :rtype: Void
=end
    def push(val)
        # Push value into stack:
        @stack.push(val)

        # Push value into minimum stack or its top if minimum didn't change:
        val = [val, (@min_stack.empty? ? val : @min_stack[-1])].min
        @min_stack.push(val) 
    end

=begin
    :rtype: Void
=end
    def pop()
        # Remove top element from both stacks:
        @stack.pop
        @min_stack.pop   
    end

=begin
    :rtype: Integer
=end
    def top()
        # Return top element of stack:
        @stack[-1]
    end

=begin
    :rtype: Integer
=end
    def get_min()
        # Return minimum element from minimum stack:
        @min_stack[-1]
    end
end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

File: 155. Min Stack: O(1).py
# 155. Min Stack
# https://leetcode.com/problems/min-stack/description/

# Stack for push, pop, top and minimum element in O(1) time:
class MinStack:
    def __init__(self):
        # One array for tracking stack and other for minimum value at each step:
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        # Push value into stack:
        self.stack.append(val)

        # Push value into minimum stack or its top if minimum didn't change:
        val = min(val, self.min_stack[-1] if self.min_stack else val)
        self.min_stack.append(val)

    def pop(self) -> None:
        # Remove top element from both stacks:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        # Return top element of stack:
        return self.stack[-1]

    def getMin(self) -> int:
        # Return minimum element from minimum stack:
        return self.min_stack[-1]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
